### DISCUSSION TITLE
Cannot define global function - undeclared reference error

### DISCUSSION THREAD

**Discussion by davenewza** [google/cel-go#1061] [Q&A]
For the expression `upper("abc") == "ABC"` I am getting this error

> ERROR: <input>:1:1: undeclared reference to 'upper' (in container '')
>         	            	 | upper
>         	            	 | ^

But I have correctly specified the function declaration, I believe?  Judging from the error, it's looking for a member function and not a global function. 

```go
env, err := cel.NewEnv(
  cel.ClearMacros(),
  cel.CustomTypeProvider(typeProvider),
  cel.Function("upper",
	  cel.Overload("string_uppercase_string",
		  []*cel.Type{cel.StringType},
		  cel.StringType,
		  cel.UnaryBinding(func(lhs ref.Val) ref.Val {
			  return types.String(strings.ToUpper(fmt.Sprintf("%s", lhs)))
		  }),
	  ),
  ),
)
```

Can someone please point me in the right direction as I am completely stuck

I am using `0.21.0`

**Comment by davenewza** [comment]
This example looks pretty much the same https://github.com/google/cel-go/blob/master/examples/custom_global_function_test.go

**Reply by TristonianJones** [reply]
Hi @davenewza,

I'm able to run the `upper` function successfully as-is though I would recommend the following naming convention for the overload: `upper_string` since it's unary and global. This is a minor comment though and won't affect functionality.

Perhaps if you could include more of the test setup?

-Tristan

**Reply by davenewza** [reply]
I figured it out, and unrelated to CEL - apologies
